File: search/array-based-search/exponential-search/cpp/exponential_search_test.cpp
#include <gtest/gtest.h>
#include <vector>

// Forward declaration
int exponentialSearch(const std::vector<int>& arr, int target);

// Test: Target found in middle
TEST(ExponentialSearchTest, TargetFoundInMiddle) {
    std::vector<int> arr = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
    EXPECT_EQ(exponentialSearch(arr, 70), 6) << "Should find 70 at index 6";
}

// Test: Target at beginning
TEST(ExponentialSearchTest, TargetAtBeginning) {
    std::vector<int> arr = {1, 2, 3, 4, 5};
    EXPECT_EQ(exponentialSearch(arr, 1), 0) << "Should find first element";
}

// Test: Target at end
TEST(ExponentialSearchTest, TargetAtEnd) {
    std::vector<int> arr = {1, 2, 3, 4, 5};
    EXPECT_EQ(exponentialSearch(arr, 5), 4) << "Should find last element";
}

// Test: Not present
TEST(ExponentialSearchTest, TargetNotFound) {
    std::vector<int> arr = {1, 2, 3, 4, 5};
    EXPECT_EQ(exponentialSearch(arr, 6), -1) << "Should return -1 for missing element";
}

// Test: Empty array
TEST(ExponentialSearchTest, EmptyArray) {
    std::vector<int> arr = {};
    EXPECT_EQ(exponentialSearch(arr, 10), -1) << "Empty array should return -1";
}

// Test: Single element found
TEST(ExponentialSearchTest, SingleElementFound) {
    std::vector<int> arr = {5};
    EXPECT_EQ(exponentialSearch(arr, 5), 0) << "Single element found → index 0";
}

// Test: Single element not found
TEST(ExponentialSearchTest, SingleElementNotFound) {
    std::vector<int> arr = {5};
    EXPECT_EQ(exponentialSearch(arr, 10), -1) << "Single element not found → -1";
}

// Test: Small array (power of two size)
TEST(ExponentialSearchTest, SmallArrayPowerOfTwo) {
    std::vector<int> arr = {1, 3, 5, 7};
    EXPECT_EQ(exponentialSearch(arr, 7), 3) << "Should find 7 at index 3";
}

// Test: Target larger than all elements
TEST(ExponentialSearchTest, TargetLargerThanAll) {
    std::vector<int> arr = {1, 2, 3, 4};
    EXPECT_EQ(exponentialSearch(arr, 10), -1) << "Should return -1 if target > max";
}

// Test: Target smaller than first element
TEST(ExponentialSearchTest, TargetSmallerThanFirst) {
    std::vector<int> arr = {10, 20, 30};
    EXPECT_EQ(exponentialSearch(arr, 5), -1) << "Should return -1 if target < min";
}

// Test: Duplicate elements — should return valid index
TEST(ExponentialSearchTest, DuplicateElementsValidIndex) {
    std::vector<int> arr = {2, 2, 2, 3, 3, 3, 4, 4};
    int result = exponentialSearch(arr, 3);
    EXPECT_GE(result, 3) << "Result must be one of the indices where 3 occurs";
    EXPECT_LE(result, 5) << "Result must be within bounds";
    EXPECT_EQ(arr[result], 3) << "Returned value must be 3";
}

// Entry point for tests
int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: search/array-based-search/exponential-search/java/ExponentialSearchTest.java
import org.junit.Test;
import static org.junit.Assert.*;

/**
 * Unit tests for the ExponentialSearch class.
 */
public class ExponentialSearchTest {

    @Test
    public void testTargetFoundInMiddle() {
        int[] arr = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
        assertEquals("Should find 70 at index 6", 6, ExponentialSearch.exponentialSearch(arr, 70));
    }

    @Test
    public void testTargetAtBeginning() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals("Should find first element", 0, ExponentialSearch.exponentialSearch(arr, 1));
    }

    @Test
    public void testTargetAtEnd() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals("Should find last element", 4, ExponentialSearch.exponentialSearch(arr, 5));
    }

    @Test
    public void testTargetNotPresent() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals("Should return -1 for missing element", -1, ExponentialSearch.exponentialSearch(arr, 6));
    }

    @Test
    public void testEmptyArray() {
        int[] arr = {};
        assertEquals("Empty array should return -1", -1, ExponentialSearch.exponentialSearch(arr, 10));
    }

    @Test
    public void testSingleElementFound() {
        int[] arr = {5};
        assertEquals("Single element found → index 0", 0, ExponentialSearch.exponentialSearch(arr, 5));
    }

    @Test
    public void testSingleElementNotFound() {
        int[] arr = {5};
        assertEquals("Single element not found → -1", -1, ExponentialSearch.exponentialSearch(arr, 10));
    }

    @Test
    public void testSmallArrayPowerOfTwo() {
        int[] arr = {1, 3, 5, 7};
        assertEquals("Should find 7 at index 3", 3, ExponentialSearch.exponentialSearch(arr, 7));
    }

    @Test
    public void testTargetLargerThanAll() {
        int[] arr = {1, 2, 3, 4};
        assertEquals("Should return -1 if target > max", -1, ExponentialSearch.exponentialSearch(arr, 10));
    }

    @Test
    public void testTargetSmallerThanFirst() {
        int[] arr = {10, 20, 30};
        assertEquals("Should return -1 if target < min", -1, ExponentialSearch.exponentialSearch(arr, 5));
    }

    @Test
    public void testDuplicates_ReturnsValidIndex() {
        int[] arr = {2, 2, 2, 3, 3, 3, 4, 4};
        int result = ExponentialSearch.exponentialSearch(arr, 3);
        assertTrue("Result must be one of the indices where 3 occurs", result >= 3 && result <= 5);
        assertEquals("Returned value must be 3", 3, arr[result]);
    }
}